File: problems/树/106.从中序与后序遍历序列构造二叉树.py
#!/usr/bin/env python3
"""
106. 从中序与后序遍历序列构造二叉树
"""
from typing import List, Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def build_tree_in_post(inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
    if not postorder: return None
    root = TreeNode(postorder[-1])
    idx = inorder.index(postorder[-1])
    root.left  = build_tree_in_post(inorder[:idx], postorder[:idx])
    root.right = build_tree_in_post(inorder[idx+1:], postorder[idx:-1])
    return root

def inorder_list(root: Optional[TreeNode]) -> List[int]:
    if not root: return []
    return inorder_list(root.left) + [root.val] + inorder_list(root.right)

if __name__ == "__main__":
    inorder1  = [9, 3, 15, 20, 7]
    postorder1 = [9, 15, 7, 20, 3]
    root1 = build_tree_in_post(inorder1, postorder1)
    print(inorder_list(root1))  # [9, 3, 15, 20, 7]
    print(inorder_list(build_tree_in_post([1], [1])))  # [1]
    print(build_tree_in_post([], []))  # None

File: problems/树/106.从中序与后序遍历序列构造二叉树.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <functional>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* buildTreeInPost(vector<int>& inorder, vector<int>& postorder) {
    unordered_map<int, int> idx;
    for (int i = 0; i < (int)inorder.size(); i++) idx[inorder[i]] = i;
    function<TreeNode*(int,int,int)> build = [&](int il, int ir, int pl) -> TreeNode* {
        if (il > ir) return nullptr;
        TreeNode* root = new TreeNode(postorder[pl]);
        int mid = idx[postorder[pl]];
        root->left  = build(il, mid-1, pl-(ir-mid)-1);
        root->right = build(mid+1, ir, pl-1);
        return root;
    };
    return build(0, (int)inorder.size()-1, (int)postorder.size()-1);
}
void inorderPrint(TreeNode* root) {
    if (!root) return;
    inorderPrint(root->left);
    cout << root->val << " ";
    inorderPrint(root->right);
}
int main() {
    vector<int> in1  = {9, 3, 15, 20, 7};
    vector<int> post1 = {9, 15, 7, 20, 3};
    inorderPrint(buildTreeInPost(in1, post1)); cout << endl;  // 9 3 15 20 7
    vector<int> in2  = {1};
    vector<int> post2 = {1};
    inorderPrint(buildTreeInPost(in2, post2)); cout << endl;  // 1
    vector<int> in3, post3;
    cout << (buildTreeInPost(in3, post3) == nullptr ? "null" : "?") << endl;  // null
    return 0;
}
